thirdparty/icu4c/common/umutablecptrie.cpp

```diff
@@ -1304,7 +1304,7 @@ int32_t MutableCodePointTrie::compactIndex(int32_t fastILimit, MixedBlocks &mixe
         errorCode = U_MEMORY_ALLOCATION_ERROR;
         return 0;
     }
-    uprv_memcpy(index16, fastIndex, fastIndexLength * 2);
+    uprv_memcpy(index16, fastIndex, index16Capacity * 2);
```

Comment:
> warning: 'memcpy' writing between 128 and 2048 bytes into a region of size 24 overflows the destination [-Wstringop-overflow=]

> `index16Capacity = fastIndexLength + index1Length + index3Capacity + index2Capacity + 1`

Seems like `index16Capacity` should be bigger than `fastIndexLength`, if it was overflowing with smaller size, I'm not sure how it will fix it. I guess the compiler assumes some added values can overflow and become negative (which probably should never happen).

I'm not familiar with this part of code but using `fastIndexLength` to copy `fastIndex` seems like a more likely thing that using target buffer size.

Reply:
Actually, seems like the reason is how `uprv_malloc` work, compiler assume `index16Capacity * 2` can be `<= 0` (which should be false), and in this case `uprv_malloc` return pointer to 24 bytes `zeroMem` instead of `nullptr`.

In any case, it is the same, it is only an issue if `index16Capacity * 2` overflow.

Reply:
I see, I misunderstood the warning and what this code is doing.

Checking the code further below, there's further copies at other positions to fill the whole buffer so indeed this first one wasn't meant to fill everything.

Reply:
> Actually, seems like the reason is how `uprv_malloc` work, compiler assume `index16Capacity * 2` can be `<= 0` (which should be false), and in this case `uprv_malloc` return pointer to 24 bytes `zeroMem` instead of `nullptr`.
>
> In any case, it is the same, it is only an issue if index16Capacity * 2 overflow.

Right, these kind of GCC int overflow warnings are really annoying to workaround. Sadly https://gcc.gnu.org/bugzilla/show_bug.cgi?id=80922 means that we can't fully ignore those warnings in thirdparty code, they still show up when linking with LTO :|